File: branches/mOrc/OrcJava/src/orc/lib/update/ReplaceNode.java
//
// ReplaceNode.java -- Java class ReplaceNode
// Project OrcJava
//
// $Id$
//
// Created by jthywiss on Mar 17, 2010.
//
// Copyright (c) 2010 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.update;

import java.util.ArrayList;
import java.util.List;

import orc.ast.oil.AstNode;
import orc.ast.oil.expression.Def;
import orc.ast.oil.expression.Expression;
import orc.runtime.Token;
import orc.runtime.Token.FrameContinuation;
import orc.runtime.values.Closure;

/**
 * Edit operation that is a to-one mapping of an AST node. 
 *
 * @author jthywiss
 */
public class ReplaceNode extends AstEditOperation {
	private final AstNode oldNode;
	private final AstNode newNode;

	/**
	 * Constructs an object of class ReplaceNode.
	 *
	 * @param oldNode
	 * @param newNode
	 */
	public ReplaceNode(final AstNode oldNode, final AstNode newNode) {
		super();
		this.oldNode = oldNode;
		this.newNode = newNode;
	}

	/* (non-Javadoc)
	 * @see orc.lib.update.AstEditOperation#isTokenAffected(orc.runtime.Token)
	 */
	@Override
	public boolean isTokenAffected(final Token token) {
		return token.getNode() == oldNode;
	}

	/* (non-Javadoc)
	 * @see orc.lib.update.AstEditOperation#isTokenSafe(orc.runtime.Token)
	 */
	@Override
	public boolean isTokenSafe(final Token token) {
		return true;
	}

	/* (non-Javadoc)
	 * @see orc.lib.update.AstEditOperation#migrateToken(orc.runtime.Token)
	 */
	@Override
	public boolean migrateToken(final Token token) {
		if (isTokenAffected(token)) {
			System.err.println(">>Move " + token + " from " + oldNode + " to " + newNode);
			token.move((Expression) newNode);
			return true;
		} else {
			return false;
		}
	}

	/* (non-Javadoc)
	 * @see orc.lib.update.AstEditOperation#migrateClosures(orc.runtime.Token)
	 */
	@Override
	protected void migrateClosures(final Token token) {
		migrateClosures(token.getEnvironment().items(), new ArrayList<Closure>());
	}

	private void migrateClosures(final List<Object> os, final List<Closure> completedClosures) {
		for (final Object o : os) {
			if (o instanceof Closure) {
				final Closure c = (Closure) o;
				if (c.def == oldNode) {
					c.def = (Def) newNode;
				}
				final List<Object> cEnv = c.env.items();
				while (cEnv.removeAll(completedClosures)) {
					// Remove any recursive bindings
				}
				completedClosures.add(c);
				migrateClosures(cEnv, completedClosures);
			}
		}
	}

	/* (non-Javadoc)
	 * @see orc.lib.update.AstEditOperation#migrateFrameStack(orc.runtime.Token)
	 */
	@Override
	protected void migrateFrameStack(final Token token) {
		if (token.getContinuation() == null) {
			return;
		}
		for (FrameContinuation c = token.getContinuation(); c != null; c = c.parent) {
			if (c.callPoint == oldNode) {
				c.callPoint = (Expression) newNode;
			}
		}
	}

}

File: OrcScala/src/orc/lib/progswap/ReplaceNode.scala
//
// ReplaceNode.scala -- Scala class ReplaceNode
// Project OrcScala
//
// $Id$
//
// Created by jthywiss on Sep 30, 2010.
//
// Copyright (c) 2011 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//
package orc.lib.progswap

import orc.ast.AST
import orc.ast.oil.nameless.{ Def, Expression, NamelessAST }
import orc.run.extensions.SwappableASTs
import orc.run.Orc
import orc.run.core.BoundValue
import orc.run.core.Token
import orc.run.core.Closure
import orc.run.core.Binding
import orc.run.core.FunctionFrame
import orc.run.core.SequenceFrame

/** Edit operation that is a to-one mapping of an AST node.
  *
  * @author jthywiss
  */
case class ReplaceNode[A <: AST, B <: AST](oldNode: A, newNode: B) extends AstEditOperation {
  def tokenCracker(token: Token): Token = token.asInstanceOf[Token]

  def isTokenAffected(token: Token): Boolean = { tokenCracker(token).getNode() == oldNode }

  def isTokenSafe(token: Token): Boolean = { true }

  def migrateToken(token: Token): Boolean = {
    if (isTokenAffected(token)) {
      Console.err.println(">>Move " + token + " from " + oldNode + " to " + newNode)
      tokenCracker(token).move(newNode.asInstanceOf[Expression])
      true
    } else {
      false
    }
  }

  def migrateClosures(token: Token) {
    migrateClosures(tokenCracker(token).getEnv(), Nil)
  }

  private def migrateClosures(bindings: List[Binding], completedClosures: List[Closure]) {
    for (binding <- bindings)
      binding match {
        case bv: BoundValue => {
          bv.v match {
            case c: Closure => {
              SwappableASTs.setClosureDef(c, c.collection.definitions map { d => if (d == oldNode) newNode.asInstanceOf[Def] else d })
              // Filter recursive bindings
              val cEnv = c.lexicalContext filterNot {
                case bv2: BoundValue => {
                  bv2.v match {
                    case c2: Closure => completedClosures.contains(c2)
                  }
                }
                case _ => false
              }
              migrateClosures(cEnv, /*c ::*/ completedClosures)
            }
            case _ => {}
          }
        }
        case _ => {}
      }
  }

  def migrateFrameStack(token: Token) {
    Console.err.println(">>ReplaceNode.migrateFrameStack(" + token + ")")
    for (frame <- tokenCracker(token).getStack()) {
      Console.err.println("  " + frame)
      frame match {
        case sf: SequenceFrame if (sf.node == oldNode) => SwappableASTs.setSequenceFrameNode(sf, newNode.asInstanceOf[Expression])
        case ff: FunctionFrame if (ff.callpoint == oldNode) => SwappableASTs.setFunctionFrameCallpoint(ff, newNode.asInstanceOf[Expression])
        case _ => {}
      }
    }
  }

}
